Theme refactor

mistBourne1300 opened the pull request:
Made it easier to add themes. Keep in mind this is v.1. I've already thought of a few things that could be handled waaaay better.

mistBourne1300 committed:
Unicode Mode and colors!

mistBourne1300 committed:
RADIO

natecraddock committed:
Use checkboxes for theme toggles

Radio buttons initially made sense, but
there wasn't a clear way to disable one of the options. We just use
checkboxes like normal, and add extra logic to disable other themes
when one is enabled.

mistBourne1300 committed:
Themes colors support
Added a check on spooky theme colors to make sure the terminal used
supports 256 colors.

mistBourne1300 committed:
check for colors with curses.can_change_colors()
actually use the function (with parentheses)

mistBourne1300 committed:
Theme Refactor
Made it easier to add themes

mistBourne1300 commented:
Issue: The header is not correctly changing the `bookend` emojis. Toggling Christmas theme, then spooky theme or birthday theme will make the Christmas emojis appear along with the spooky/birthday ones. It also happens if Christmas theme is turned on after spooky or birthday. However, toggling spooky and birthday alone do not present the problem.

Issue: Settings are not saved in preferences for which theme is toggled, so that every time zygrader is opened it opens in default, even if another theme was set before.

natecraddock commented on zygrader/config/versioning.py:
```diff
@@ -97,6 +97,14 @@ def versioning_update_preferences():
     if compare_versions(version, user_version):
         config["spooky_mode"] = False
 
+    version = "5.0.0"
+    if compare_versions(version, user_version):
+        config["unicode_mode"] = False
+
+    version = "5.0.1"
+    if compare_versions(version, user_version):
+        config["theme"] = "Default"
+
```
You're missing some changes from master here `update_config_with_new_preferences` which handles adding new preference keys automatically.

natecraddock commented on zygrader/user.py:
```diff
@@ -199,16 +203,34 @@ def is_toggled(self, _id: str):
         return preferences.get(self.__preference) == _id
 
 
+"""class ThemeRadioGroup(ui.layers.RadioGroup):
```
Looks like this isn't needed

natecraddock commented on zygrader/config/preferences.py:
```diff
@@ -28,15 +28,22 @@
     "data_dir": "",
 }
 
+THEMES = {
```
These aren't used anywhere in this file, but they are used in `themes.py`. I think it would be more logical to store this data there.

natecraddock commented on zygrader/ui/themes.py:
```diff
@@ -0,0 +1,109 @@
+    def __init__(self):
+        self.__init_colors()
+
+        self.THEME_COLORS = {
```
Usually `UPPER_CASE` is reserved for static members of the class, which are defined outside functions. See the `Event` class for examples.

I would make these static since they wouldn't change

natecraddock commented on zygrader/ui/window.py:
```diff
@@ -66,7 +68,7 @@ def __init_curses(self, stdscr, callback, args):
         curses.curs_set(0)
 
-        self.__init_colors()
+        #self.__init_colors()
```
Should be removed

natecraddock reviewed (changes requested):
Just doing a quick first pass here. When I ran this I got a crash because `unicode_mode` wasn't in my preferences. It looks like you need to merge master into your branch.

mistBourne1300 committed:
Merge branch 'master' into theme_refactor

mistBourne1300 committed:
Streamlining
streamlines the code a bit

natecraddock commented on zygrader/config/versioning.py:
```diff
@@ -107,6 +107,14 @@ def versioning_update_preferences():
     if compare_versions(version, user_version):
         config["spooky_mode"] = False
 
+    version = "5.0.0"
+    if compare_versions(version, user_version):
+        config["unicode_mode"] = False
+
+    version = "5.0.1"
+    if compare_versions(version, user_version):
+        config["theme"] = "Default"
+
```
You can delete this.

Editing "old" versioning code doesn't work. For example, most (hopefully all) TAs have used version 5.0.0 at this point. So this if statement wouldn't run for any of those students because they already have updated to that version.

Additionally, I wrote `update_config_with_new_preferences` which does this automatically for new preference variables.

mistBourne1300 replied:
Should I delete the versioning for 5.0.0 as well? All of it, goes then?

natecraddock replied:
It's safe to remove all of it. `update_config_with_new_preferences` will always add new defaults

natecraddock commented on zygrader/ui/themes.py:
```diff
@@ -0,0 +1,114 @@
+    def __init__(self):
+        self.__init_colors()
+        # cannot decalre this outside the function, since it requires curses to be initialized
+        self.THEME_COLORS = {
```
That makes sense. In this case, I would prefer `lower_snake_case` since these aren't static class data.

natecraddock replied:
Also, "decalre" typo in the comment

natecraddock commented on zygrader/ui/window.py:
```diff
@@ -54,6 +55,7 @@ def __init__(self, callback, window_name, args):
     def __init_curses(self, stdscr, callback, args):
         """Configure basic curses settings"""
         self.stdscr = stdscr
+        self.Window_Theme = themes.Theme()
```
Use `self.window_theme` for this variable

natecraddock replied:
I think `alt+F2` in vscode can rename all usages

natecraddock commented on zygrader/ui/window.py:
```diff
@@ -108,9 +93,28 @@ def __resize_terminal(self):
         if self.dark_mode:
-            return curses.color_pair(5), curses.color_pair(6)
-        return curses.color_pair(3), curses.color_pair(4)
+            themeString = self.theme.lower() + "_dark"
```
Again, snake_case here

natecraddock commented on zygrader/ui/window.py:
```diff
@@ -120,10 +124,7 @@ def __get_email_text(self):
-        if self.spooky_mode:
-            separator = "🎃"
-        self.header.erase()
+        separator = self.get_header_separator()
```
I just realized that we are missing a `self.header.erase()` before redraw. This is leading to some "ghosting" where emoji aren't being erased on switching themes 
![image](https://user-images.githubusercontent.com/7967463/98565854-55b98a00-226b-11eb-871a-2f11f4dded68.png)

natecraddock reviewed (changes requested):
Looks good! Just a few small things, mostly style issues and typos in comments.

natecraddock commented on zygrader/ui/themes.py:
```diff
@@ -0,0 +1,114 @@
+    def get_bookends(self, key: str):
+        if key not in self.THEME_BOOKENDS:
+            raise KeyError("Invalid Bookend Key")
+        return self.THEME_BOOKENDS[key]
```
End the file with a newline (convention)

natecraddock commented:
The reason why the preferences appear to not be saved is your versioning code for 5.0.1. I already mentioned it can be removed, but what happens is each time zygrader starts up it sees that the current version (5.0.0) is less thatn 5.0.1 so it resets the theme to default.

natecraddock commented on zygrader/ui/themes.py:
```diff
@@ -0,0 +1,114 @@
+        # curses can change the colors for Christmas theme, even if the terminal doesn't support the other ones.
+        # hence all this logic...
+        if not curses.can_change_color():
+            index = key.find("_")
+            theme = key[:index]
+            if theme is not "christmas":
```
Simplify to `if "christmas" not in key`

mistBourne1300 committed:
Final theme update
Fixed the issues with drawing the header and saving preferences
Also added a thanksgiving theme

natecraddock commented:
I just tested on Mate terminal (one of the problematic ones) and I cannot see the christmas colors at all.

mistBourne1300 committed:
Versioning code
quick fix to the versioning code

mistBourne1300 committed:
Christmas color change
Christmas colors (should) now be supported, even if the terminal
cannot support the other themes.
